Transaction flow

jbrauck-unchained opened the pull request:
<!--
Thank you for contributing to Caravan! Please provide the following details to help us review your pull request efficiently.
-->

**What kind of change does this PR introduce?**

This introduces a new tranasction flow diagram feature

**Issue Number:**

Fixes #393 https://github.com/caravan-bitcoin/caravan/issues/393

**Snapshots/Videos:**

<img width="3432" height="1716" alt="Screenshot 2025-10-05 at 10 54 04 PM" src="https://github.com/user-attachments/assets/8b79b0af-6b83-415d-a51d-6d29ca28a229" />



**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

I like what sparrow does to try and give a user a sense of understanding their own transaction. I think this is a small lift to gain feature parity in a place that makes caravan more educational.

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

## Checklist
- [ ] I have tested my changes thoroughly.
- [ ] I have added or updated tests to cover my changes (if applicable).
- [ ] I have verified that test coverage meets or exceeds 95% (if applicable).
- [ ] I have run the test suite locally, and all tests pass.
- [ ] I have written tests for all new changes/features
- [ ] I have followed the project's coding style and conventions.
- [ ] I have created a changeset to document my changes (`` npm run changeset ``)




**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/caravan-bitcoin/caravan/blob/main/apps/coordinator/CONTRIBUTING.md)?**

<!--Yes or No-->

**For information on creating and using changesets, please refer to our [documentation on changesets](https://github.com/caravan-bitcoin/caravan?tab=readme-ov-file#changesets)**.

jbrauck-unchained committed:
i like it so far

jbrauck-unchained committed:
I like this so far

jbrauck-unchained committed:
stopping point before clean up

jbrauck-unchained committed:
deleted the markdown slop

Commit:
Merge branch 'main' into transaction-flow

jbrauck-unchained committed:
keeps transaction preview persistent

jbrauck-unchained committed:
added high i/o handling with drawer

jbrauck-unchained committed:
Merge remote-tracking branch 'origin/main' into transaction-flow
getting on main

jbrauck-unchained committed:
Merge branch 'transaction-flow' of https://github.com/jbrauck-unchained/caravan into transaction-flow
trying to fix this

jbrauck-unchained committed:
lint and prettier fixes

jbrauck-unchained committed:
remove comment

jbrauck-unchained committed:
Merge branch 'main' into transaction-flow

jbrauck-unchained committed:
added ~ for close to 0% of transaction fees

jbrauck-unchained committed:
Merge branch 'main' into transaction-flow

jbrauck-unchained committed:
Merge branch 'main' into transaction-flow

jbrauck-unchained committed:
merge in main

jbrauck-unchained committed:
feat: add flow diagram support on txhistory

Legend101Zz commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram.tsx:
```diff
@@ -0,0 +1,1703 @@
+  }, [inputs, outputs, fee, changeAddress, inputsTotalSats]);
+
+  // Build smooth cubic-bezier path from (x1,y1) to (x2,y2)
+  const buildCurvePath = (x1: number, y1: number, x2: number, y2: number) => {
```
ok so this component seems to getting quite big , can we brak this down all into a TransactionPreview folder/directory and then have a utils , hooks and other components ( broken from this one ) there so that this file becomes manageable?

Legend101Zz commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram.tsx:
```diff
@@ -0,0 +1,1703 @@
+  };
+
+  // Format script type for display
+  const formatScriptType = (scriptType?: string) => {
```
can you add this and formatAddress in the general utils file as they can be used by other components too :)

Legend101Zz reviewed (changes requested):
@jbrauck-unchained thanks for the awesome work , I tested it and it looks super good to me , now I guess we can move towards the code changes , the main concern I have is over modularity and an IIFE used in rendering of the flow status

<img width="1177" height="814" alt="Screenshot 2026-01-06 at 07 22 52" src="https://github.com/user-attachments/assets/5ff403fa-c133-425d-b6ca-63f36b9c046c" />

jbrauck-unchained committed:
fix: resolve merge conflict comment and refactor transaction status calculation

jbrauck-unchained committed:
refactor: break down TransactionFlowDiagram into modular components

- Create TransactionFlowDiagram directory structure with separate files:
  - index.tsx: Main component file (cleaner, more focused)
  - hooks.ts: Custom useFlowPaths hook for SVG path calculations
  - utils.ts: Utility functions (buildCurvePath, formatAddress, formatScriptType, getScriptTypeColor, getStatusDisplay)
  - FlowDrawers.tsx: Input and output drawer components
  - FlowSummary.tsx: Summary section with legend and transaction totals
- Move formatSats and formatAddress to transactionFlowUtils.ts for reuse
- Remove old 1704-line monolithic TransactionFlowDiagram.tsx file

This addresses code review feedback to improve maintainability and readability.

jbrauck-unchained committed:
fix: resolve linting errors from refactoring

- Remove unused variables in TransactionPreview.jsx render method
- Fix prettier formatting in TransactionsTable.tsx
- Fix prettier formatting in useTransactionDetails.ts
- Remove unused blockHeight variable in transactionFlowUtils.ts

jbrauck-unchained committed:
Merge branch 'main' into transaction-flow

Legend101Zz reviewed (approved):
Thanks alot for the changes , it's looking really good now , nothing from my end now :) 

@bucko13 can you please review it when you get the chance ...

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/FlowDrawers.tsx:
```diff
@@ -0,0 +1,356 @@
+          }}
+        >
+          {inputs.map((input, idx) => {
+            const inputAmount = BigNumber(
```
We should pull this out to be its own component and then just pass in the props.

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/FlowSummary.tsx:
```diff
@@ -0,0 +1,350 @@
+              label="Economical"
+              color="success"
+              size="small"
+              sx={{ height: 20, fontSize: "0.7rem" }}
```
seems like there's some room for code consolidation here. this styling is repeated on every box as far as I can tell and it should be generalized. If we change the look somehow you'll have to change every single line. do we want this to be a specialized kind of component?

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/index.tsx:
```diff
@@ -0,0 +1,976 @@
+            }}
+          >
+            {inputs.slice(0, 4).map((input, idx) => {
+              const inputAmount = BigNumber(
```
another one that should be its own component to simplify this file a bit. and what can we do to clean up the inline styles?

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/index.tsx:
```diff
@@ -0,0 +1,976 @@
+
+          {/* Change Outputs */}
+          {flowData.changeOutputs.map((output, idx) => {
+            const amount = BigNumber(output.amount);
```
modular component opportunity here.

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/index.tsx:
lots of simplification we can get here.

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/utils.ts:
```diff
@@ -0,0 +1,104 @@
+  if (!theme) return "#9e9e9e"; // fallback grey
+
+  switch (scriptType?.toLowerCase()) {
+    case "p2wsh":
```
we have constants exported from caravan-bitcoin that we can use to standardize on for the script types. 

also note for outputs we can have a p2tr output

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionFlowDiagram/utils.ts:
```diff
@@ -0,0 +1,104 @@
+  if (!theme) return { label: "Unknown", color: "#9e9e9e" }; // fallback
+
+  switch (status) {
+    case "draft":
```
status would be nice to have as a type-safe enum

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionsTab/TableComponents/PendingTransactionsView.tsx:
```diff
@@ -90,13 +93,16 @@ export const PendingTransactionsView: React.FC<Props> = ({
             onClickTransaction={onClickTransaction}
             onAccelerateTransaction={onAccelerateTransaction}
             showAcceleration={true}
+            expandable={true}
```
cleaner way to do this is
```suggestion
            expandable
```

but also why do we need an option here if it's always expandable? are there some views where we don't expand it?

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionsTab/TableComponents/TransactionsTable.tsx:
```diff
@@ -252,151 +267,310 @@ const TransactionTableHeader: React.FC<{
+  isExpanded,
+  onToggleExpand,
   network,
+  walletAddresses = [],
```
the component above uses a hook to get the addresses just to pass them down here. this component is then also using hooks to populate its internal state. Why not just have the hook for the wallet addresses here and simplify the function signature?

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionsTab/TableComponents/TransactionsTable.tsx:
```diff
@@ -422,12 +596,18 @@ export const TransactionTable: React.FC<TransactionTableProps> = ({
+                  isExpanded={expandedTxid === tx.txid}
+                  onToggleExpand={() => handleToggleExpand(tx.txid)}
                   network={network}
+                  walletAddresses={walletAddresses}
```
actually we're prop drilling, passed through multiple levels. definitely doesn't need to do that.

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionPreview.jsx:
```diff
@@ -193,34 +184,38 @@ class TransactionPreview extends React.Component {
     downloadFile(psbtData, "transaction.psbt");
   }
 
+  getTransactionStatus() {
```
seems like a nice, standalone utility function. could make it type safe too so we know exactly what we expect it to return.

bucko13 commented on apps/coordinator/src/components/Wallet/TransactionPreview.jsx:
```diff
@@ -193,34 +184,38 @@ class TransactionPreview extends React.Component {
+    const { signatureImporters, requiredSigners, broadcasting, txid } =
+      this.props;
+
+    if (broadcasting) return "broadcast-pending";
```
why is `broadcasting` translate to `broadcast-pending`? is this actually that it's in the mempool? if so "broadcast pending" isn't correct.

bucko13 replied:
if not, maybe we call it "authoring"? if "Broadcasting" being true indicates it's being authored then I think this variable is mislabeled.

bucko13 commented on apps/coordinator/src/hooks/useTransactionDetails.ts:
```diff
@@ -0,0 +1,100 @@
+      if (blockchainClient.type === "public") {
+        try {
+          // Fetch raw transaction from mempool.space which includes prevout
+          const response = await blockchainClient.Get(`/tx/${txid}`);
```
something funky going on here. the whole point of the blockchain client and having its own package `@caravan/clients` is to add an abstraction layer so that this type of code doesn't need to be written. If the package supports these requests, then use the existing methods. if not, then let's add support there.

bucko13 commented on apps/coordinator/src/hooks/useTransactionDetails.ts:
```diff
@@ -0,0 +1,100 @@
+  const blockchainClient = useGetClient();
+
+  return useQuery(
+    [TRANSACTION_DETAILS_KEY, txid],
```
see some of the other hooks for how we setup the query cache keys.

bucko13 commented on apps/coordinator/src/reducers/transactionReducer.js:
```diff
@@ -455,6 +457,8 @@ export default (state = initialState(), action) => {
       return updateState(state);
     case SET_TXID:
       return updateState(state, { txid: action.value });
+    case SET_BROADCASTING:
```
now i'm curious what this is actually indicating. also wonder if there's a way we can avoid adding to the redux store. It's hard in caravan since we have a lot of tightly coupled legacy code, but if we can avoid adding to the redux store that would definitely be the preference.

bucko13 commented on apps/coordinator/src/utils/transactionFlowUtils.ts:
```diff
@@ -0,0 +1,266 @@
+  changeAddress?: string;
+  inputsTotalSats: BigNumber;
+  status:
+    | "draft"
```
see we're using these values in other places. what if we made this a shareable type and then could make all uses of it type safe.

bucko13 commented on apps/coordinator/src/utils/transactionFlowUtils.ts:
```diff
@@ -0,0 +1,266 @@
+/**
+ * Detect script type from address prefix (heuristic)
+ */
+const detectScriptTypeFromAddress = (address?: string): string | undefined => {
```
we have address type detection tools in caravan-bitcoin

bucko13 commented on apps/coordinator/src/utils/transactionFlowUtils.ts:
```diff
@@ -0,0 +1,266 @@
+ * @returns FlowDiagramProps ready for the component
+ */
+export const transformTransactionToFlowDiagram = (
+  tx: any,
```
I think we have some types for this don't we @Legend101Zz ?

bucko13 reviewed (commented):
I don't love all of the inline stylings that we have throughout this. it will make the aesthetics come out inconsistent with the rest of the app and also harder to maintain. we should be able to use existing theming or components and/or update to the theme.

few other opportunities for code consolidation/cleanup. Also would like to add type safety where possible.

jevidon commented:
Overall really solid. Just a couple nits I noticed, but not blocking.

<img width="1239" height="951" alt="image" src="https://github.com/user-attachments/assets/3d804fff-fab8-44de-b934-016a891d1aa9" />
